crates/sui-graphql-rpc/src/types/coin.rs

```diff
@@ -290,34 +293,67 @@ impl Coin {
         page: Page<object::Cursor>,
         coin_type: TypeTag,
         owner: Option<SuiAddress>,
-        checkpoint_sequence_number: Option<u64>,
+        checkpoint_viewed_at: Option<u64>,
     ) -> Result<Connection<String, Coin>, Error> {
-        let (prev, next, results) = db
-            .execute(move |conn| {
-                page.paginate_query::<StoredObject, _, _, _>(conn, move || {
-                    use objects::dsl;
-                    let mut query = dsl::objects.into_boxed();
-
-                    query = query.filter(
-                        dsl::coin_type.eq(coin_type.to_canonical_string(/* with_prefix */ true)),
-                    );
-
-                    if let Some(owner) = &owner {
-                        // Leverage index on objects table
-                        query = query.filter(dsl::owner_type.eq(OwnerType::Address as i16));
-                        query = query.filter(dsl::owner_id.eq(owner.into_vec()));
-                    }
+        let mut conn: Connection<String, Coin> = Connection::new(false, false);
+
+        let checkpoint_viewed_at =
+            match validate_cursor_consistency(checkpoint_viewed_at, page.after(), page.before()) {
+                Ok(checkpoint_viewed_at) => checkpoint_viewed_at,
+                Err(e) => return Err(e),
+            };
+
+        let response = db
+            .execute_repeatable(move |conn| {
+                let (lhs, mut rhs) = Checkpoint::available_range(conn)?;
 
-                    query
-                })
+                if let Some(checkpoint_viewed_at) = checkpoint_viewed_at {
+                    if checkpoint_viewed_at < lhs || rhs < checkpoint_viewed_at {
+                        return Ok::<_, diesel::result::Error>(None);
+                    }
+                    rhs = checkpoint_viewed_at;
+                }
+
+                let result = page.paginate_raw_query::<StoredHistoryObject, _>(
+                    conn,
+                    move |element| {
+                        element.map(|obj| {
+                            let cursor = Cursor::new(HistoricalObjectCursor::new(
+                                obj.object_id.clone(),
+                                rhs as u64,
+                            ));
+                            cursor
+                        })
+                    },
+                    move || {
+                        let final_ = Object::build_raw_consistent_query(
+                            |query| Self::filter(query, coin_type.clone(), owner),
+                            lhs as i64,
+                            rhs as i64,
+                        );
+
+                        filter!(final_, "newer.object_version IS NULL")
```

Comment:
This is a little bit nasty -- I would suggest that you bake this into the notion of a "consistent query" and check for an `ObjectFilter` that includes `object_keys` earlier on, so that you know in that case that you don't want to perform a consistent query, you just want to perform a historical query (which you could maybe factor out as well).